selfdrive/locationd/magd.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
+import os
+import sys
+import signal
+import numpy as np
+from typing import Any, Optional, Tuple, Union
+
+import cereal.messaging as messaging
+from cereal import car, log
+from openpilot.common.params import Params
+from openpilot.common.realtime import config_realtime_process
+from openpilot.common.filter_simple import FirstOrderFilter
+from openpilot.system.swaglog import cloudlog
+from openpilot.selfdrive.locationd.helpers import PointBuckets
+
+POINTS_PER_BUCKET = 3000
+MIN_POINTS_TOTAL = 4000
+FIT_POINTS_TOTAL = 4000
+MIN_VEL = 5  # m/s
+FILTER_DECAY = 0.1
+FILTER_DT = 0.015
+BUCKET_KEYS = [(-np.pi, -np.pi / 2), (-np.pi / 2, 0), (0, np.pi / 2), (np.pi / 2, np.pi)]
+MIN_BUCKET_POINTS = 1000
+CALIB_DEFAULTS = np.array([0.0, 0.0, 1.0, 1.0, 0.0, 0.0])
+VERSION = 1  # bump this to invalidate old parameter caches
+
+
+class MagBuckets(PointBuckets):
+  def add_point(self, x: float, y: float, yaw: float) -> None:
+    for bound_min, bound_max in self.x_bounds:
+      if (yaw >= bound_min) and (yaw < bound_max):
+        self.buckets[(bound_min, bound_max)].append([x, y, yaw])
+        break
+
+
+class MagCalibrator:
+  def __init__(self, CP: car.CarParams) -> None:
+    self.reset()
+
+    # try to restore cached params
+    params = Params()
+    params_cache = params.get("MagnetometerCarParams")
```

Comment:
paramsd, torqued, and magd also use this pattern. can we pull this out? then we can also share the carParams they store